### PR TITLE
A way to log which passes are run and how long they take

### LINKED ISSUE #2259: PassManager should record execution time for each pass
### What is the expected enhancement?
The `PassManager` should record the execution time for each pass that it runs, and this information should be easily retrievable.

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #2259 

```python
passmanager = PassManager() 
passmanager.append(PassC_TP_RA_PA()) 
passmanager.append(PassB_TP_RA_PA())                                                                                                      

passmanager.log_passes = True  # <- activates logging!

transpile(QuantumCircuit(QuantumRegister(1)), pass_manager=passmanager)                                                                           
passmanager.property_set['pass_log']                                                                                                              
```

```
[{'end_time': 1558122318.479962,
  'name': 'PassA_TP_NR_NP',
  'running_time': 2.5033950805664062e-05,
  'start_time': 1558122318.479937},
 {'end_time': 1558122318.479984,
  'name': 'PassC_TP_RA_PA',
  'running_time': 7.152557373046875e-06,
  'start_time': 1558122318.479977},
 {'end_time': 1558122318.480004,
  'name': 'PassB_TP_RA_PA',
  'running_time': 5.0067901611328125e-06,
  'start_time': 1558122318.479999}]
```

### CODE REVIEW

**Comment by ajavadia** [comment]
The start_time doesn't add much information, so i think it's better to remove that.

**Comment by 1ucian0** [comment]
I thought it could be handy in some debugging situations. For example, to calculate the time between passes...

But I can remove `end_time` and `start_time` if you prefer...

**Comment by 1ucian0** [comment]
I thought it could be handy in some debugging situations. For example, to calculate the time between passes...

But I can remove `end_time` and `start_time` if you prefer...

**Comment by ajavadia** [comment]
They kind of pollute the output in my opinion, especially as they are represented as `1558122318.479962`! We just want to see which passes took how long.

Also is the unit of `running_time` seconds? So each pass here ran in a few microseconds? (That seems a bit too fast?). I think we should normalize the time to milliseconds for easier reading and write the unit in parentheses (e.g. `running_time (ms)`).

Also the logging shouldn't have much overhead right? So why not enable it by default? It saves an extra command.

File: test/python/transpiler/test_pass_scheduler.py [hunk 1]
```diff
@@ -600,8 +600,83 @@ def test_conditional_and_loop(self):
         self.assertEqual(expected, passmanager.passes())
 
 
+class TestLogPasses(QiskitTestCase):
+    """ Testing the log_passes option """
```
Comment (ajavadia) [inline test/python/transpiler/test_pass_scheduler.py]: For code style and consistency's sake, can you remove all the extra spaces at the beginning and end of docstrings? https://www.python.org/dev/peps/pep-0257/
```suggestion
    """Testing the log_passes option."""
```

**Comment by 1ucian0** [comment]
> They kind of pollute the output in my opinion, especially as they are represented as 1558122318.479962! We just want to see which passes took how long.

They are "seconds since the Epoch". It's a fairly abstract and standard way to represent a point in time.

```
>>> time.ctime(1558122318.479962)
'Fri May 17 15:45:18 2019'
```

I think is better to separate the data collection from the data presentation. 

> Also is the unit of running_time seconds?

yes.

> So each pass here ran in a few microseconds? (That seems a bit too fast?).

The example is with dummy passes. But yes, passes run in fraction of seconds. Here you have a more realistic-ish example:

```python
circuit = QuantumCircuit(qr)
circuit.h(qr[0])
circuit.h(qr[0])
circuit.cx(qr[0], qr[1])
circuit.cx(qr[1], qr[0])
circuit.cx(qr[0], qr[1])
circuit.cx(qr[1], qr[0])

coupling_map = CouplingMap([[1, 0]])
basis_gates = ['u1', 'u2', 'u3', 'cx', 'id']

pm = default_pass_manager(TranspileConfig(basis_gates=basis_gates,
                                          coupling_map=coupling_map,
                                          optimization_level=3,
                                          initial_layout=None,
                                          seed_transpiler=None))
pm.log_passes = True
transpile(circuit, pass_manager=pm)

[ print("%s: %f10" % (i['name'], i['running_time'])) for i in pm.property_set['pass_log']]
```
```
SetLayout: 0.00001110
Unroller: 0.00171210
TrivialLayout: 0.00009610
CheckMap: 0.00074810
FullAncillaAllocation: 0.00002610
EnlargeWithAncilla: 0.00001810
Unroll3qOrMore: 0.00004510
BarrierBeforeFinalMeasurements: 0.00002110
LegacySwap: 0.01157910
Decompose: 0.00002610
CXDirection: 0.00866610
Optimize1qGates: 0.00864210
CXCancellation: 0.00045110
RemoveResetInZeroState: 0.00002010
Depth: 0.00130910
FixedPoint: 0.00001910
Optimize1qGates: 0.00505910
CXCancellation: 0.00041210
RemoveResetInZeroState: 0.00002910
Depth: 0.00064510
FixedPoint: 0.00004110
```

> I think we should normalize the time to milliseconds for easier reading and write the unit in parentheses (e.g. running_time (ms)).

That is part of the data presentation. It should be up to the user.

> Also the logging shouldn't have much overhead right? So why not enable it by default? It saves an extra command.

Most of the runs (by far), do not need this data, right?

**Comment by ajavadia** [comment]
I know you can write the following to get it formatted:
```
print("%s: %f10 (ms)" % (i['name'], i['running_time']*1000)) for i in pm.property_set['pass_log']
```
but this is quite cumbersome. 

Standard representations are in principle the correct way to go about things, but one shouldn't blindly apply that principle. Here we are logging transpiler passes. Passes do not take seconds, they take milliseconds or less. Also I can't think of a scenario where 15 digits of precision is needed. 

To me it comes down to what this function is for. I anticipate this is mostly used to log and read by a human. So human-readability would be nice.

**Comment by 1ucian0** [comment]
After 932940da27a96cf2a28fdf9a121f4e72aca33548:
```python
for i in pm.property_set['pass_log']:
    print(i['formatted_time'])
```
```
SetLayout: 0.00525 (ms)
Unroller: 0.88811 (ms)
TrivialLayout: 0.04506 (ms)
CheckMap: 0.35572 (ms)
FullAncillaAllocation: 0.02313 (ms)
...
```

**Comment by ajavadia** [comment]
Thanks this is much better now. 

I think instead of "formatted_time" being a separate field for each of the passes, it could be a full dictionary in property_set. The idea being that you keep `property_set['pass_log']` a machine-readable/pure log, and then expose `property_set['simple_pass_log']` (slight name change) for human-readability. What do you think?

**Comment by 1ucian0** [comment]
I'm moving the "pure" log to `pass_raw_log` and setting `pass_log` to the formatted log:
```
for i in pm.property_set['pass_log']:
    print(i)
```

**Pull request merged**